src/models/binary_graph_operations.py

```diff
@@ -0,0 +1,118 @@
+"""Contains tools for binary operations between GeoGraph objects"""
+from typing import List
+from numpy import ndarray
+from src.models.polygon_utils import (
+    connect_with_interior,
+    connect_with_interior_or_edge,
+    connect_with_interior_or_edge_or_corner,
+    connect_with_interior_bulk,
+    connect_with_interior_or_edge_bulk,
+    connect_with_interior_or_edge_or_corner_bulk,
+)
+
+# For switching identifiction mode in `identify_node`
+_SPATIAL_IDENTIFICATION_FUNCTION = {
+    "corner": connect_with_interior_or_edge_or_corner,
+    "edge": connect_with_interior_or_edge,
+    "interior": connect_with_interior,
+}
+_BULK_SPATIAL_IDENTIFICATION_FUNCTION = {
+    "corner": connect_with_interior_or_edge_or_corner_bulk,
+    "edge": connect_with_interior_or_edge_bulk,
+    "interior": connect_with_interior_bulk,
+}
+
+
+def identify_node(node: dict, other_graph: "GeoGraph", mode: str = "corner") -> ndarray:
+    """
+    Return list of all node ids in `other_graph` which identify with the given `node`.
+
+    Args:
+        node (dict): The node for which to find nodes in `other_graphs` that can be
+            identified with `node`.
+        other_graph (GeoGraph): The GeoGraph object in which to search for
+            identifications
+        mode (str, optional): Must be one of `corner`, `edge` or `interior`. Defaults
+            to "corner".
+            The different modes correspond to different rules for identification:
+            - corner: Polygons of the same `class_label` which overlap, touch in their
+                edges or corners will be identified with each other. (fastest)
+            - edge: Polygons of the same `class_label` which overlap or touch in their
+                edges will be identified with each other.
+            - interior: Polygons of the same `class_label` which overlap will be
+                identified with each other. Touching corners or edges are not counted.
```

Comment:
Question out of interest: are there examples where anything apart from using the `interior` mode would make sense in our case? Maybe, noise and inaccuracy in CRS projection could warrant the others? @Croydon-Brixton: you probably thought more about this, would be curious to hear what examples you were thinking about.

Reply:
Good question! 

My thoughts at the moment are:
- `interior` makes most sense when we stack graphs on top of each other along the time-axis.
- `edge` makes most sense when we want to combine two graphs which overlap/touch in a certain region to one larger graph (I was thinking scalablility here, maybe if we have to distribute different graphs on different workers on a cluster or simply if we have run an individual analysis in region A and one in region B before and want an easier way to get A+B than rerunning it on the A+B composite image by simply merging the graphs spatially). It would essentially catch "touching" polygons like in this image.
![image](https://user-images.githubusercontent.com/19754061/109787517-38d40880-7c06-11eb-9eb5-ae8d6f54f664.png)
- `corner`: I am not yet sure in which use case we would want to do that. Potentially when working on very high-res data and small features, where maybe a river is only diagonally connected in some cases this could be useful to combine the  individual river nodes into one larger river node, despite them lying diagonally, by going over all river-type nodes and corner-identifying them with their neighbors. This could reduce the 1-pixel noise artefacts that come from rasterization.

Reply:
Good ideas - but do you know if the rasterisation ever produces setups like in your `edge` diagram? If there are two pixels of the same class touching each other on a full pixel edge, it should just merge them right?

Reply:
Correct, in the `edge` case the polygonisation will merge this automatically, so `edge` is really only needed when users cut the raster before, polygonise them separately and then want to merge them back again.

So, in summary, the three main applications I see at the moment are:
- `interior`: for time-stacking
- `edge`: for glueing together different graphs (created by separate polygonisation
- `corner`: for fixing diagonally connected nodes